### PR TITLE
feat(data-designer): report job token usage

### PR DESCRIPTION
**Pull request by mckornfield**
## Summary

Teach Data Designer create jobs to aggregate the library's per-request token events and report cumulative input/output totals through `JobContext.usage`. This is stacked on #2107, which introduces the shared reporter contract.

## Changes

- Capture Data Designer token events under a unique run correlation.
- Sum successful provider calls, including retries, and ignore events from unrelated concurrent runs.
- Publish once when generation exits, including partial usage before a generation failure propagates.
- Preserve unknown usage when no token event is emitted.
- Add tests for totals, run isolation, empty runs, failure handling, and correlation restoration.

## Type of Change

- [x] Code change (feature, bug fix, or refactor)
- [ ] Code change with documentation updates
- [ ] Documentation only
- [ ] Contributor tooling or automation
- [ ] CI, build, or test infrastructure

## Quality Gates

- [x] Tests added or updated for changed behavior
- [ ] Existing tests cover changed behavior — justification:
- [ ] Tests not applicable — justification:
- [ ] Documentation updated for user-visible behavior
- [x] Documentation not applicable — justification: The shared contract and producer roadmap are documented in the parent PR; this change implements that documented Data Designer behavior.

## Verification

- [x] Pull request title follows the repository's Conventional Commit format
- [x] Every commit includes an appropriate `Signed-off-by:` trailer
- [x] `uv run pre-commit run -a` passes, or any blocked checks are identified below
- [x] Targeted tests pass, or tests are marked not applicable above
- [x] No secrets, API keys, or credentials are included

Targeted validation:

- `uv run --frozen --package nemo-data-designer-plugin pytest plugins/nemo-data-designer/tests/unit -v` — 111 passed.
- `uv run --frozen ty check` on changed Python files — passed.
- `uv run --frozen ruff check` and `uv run --frozen ruff format --check` on changed Python files — passed.
- `uv run pre-commit run -a` — passed.

## Summary by CodeRabbit

* **New Features**
  * Dataset creation now records input and output token usage in job usage details.
  * Token usage is aggregated for the active operation and reported when available.

* **Bug Fixes**
  * Usage tracking remains accurate across nested operations and failures.

* **Tests**
  * Added coverage for aggregation, run filtering, empty usage, failure handling, and correlation restoration.

### CODE REVIEW

File: plugins/nemo-data-designer/src/nemo_data_designer_plugin/jobs/run.py [hunk 1]
```diff
@@ -80,11 +81,12 @@ def _on_batch_complete(_: Path) -> None:
         dd_ctx=dd_ctx,
     )
     data_designer.set_run_config(dd.RunConfig(buffer_size=BUFFER_SIZE))
-    dataset_creation_results = data_designer.create(
-        config_builder,
-        num_records=step_config.job_config.num_records,
-        on_batch_complete=_on_batch_complete,
-    )
+    with capture_token_usage(ctx.usage):
```
Comment (mckornfield) [inline plugins/nemo-data-designer/src/nemo_data_designer_plugin/jobs/run.py:84]: interesting aspect way, why couldn't we just plug into `_on_batch_complete`

Comment (tylersbray) [reply plugins/nemo-data-designer/src/nemo_data_designer_plugin/jobs/run.py]: It's kind of nice, the capture_token_usage just does that yield and has the finally cleanup.  This reads well and feels ordered well, having an extra line calling some other function might look more cluttered and also maybe risk dropping things if this dd call itself blows up?

My (edited) comment is that I looked ahead and see similar for some of the other plugins.  Might be worth asking the agent if the pattern can be abstracted out?  I wonder if one risk of the stacked PR paradigm is that you risk code denormalization?

File: plugins/nemo-data-designer/tests/unit/test_token_usage.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+def test_capture_token_usage_leaves_no_report_without_events() -> None:
+    reporter = LocalJobUsageReporter()
+
+    with capture_token_usage(reporter):
+        pass
+
+    assert reporter.latest is None
+
+
+def test_capture_token_usage_reports_before_propagating_failure() -> None:
```
Comment (tylersbray) [inline plugins/nemo-data-designer/tests/unit/test_token_usage.py]: Yeah, here and below the tests hint at the wrapping approach actually making for good insurance.

**Review by tylersbray — APPROVED** [review]
Approving, note all these approvals are modulo CI passing, code rabbit being happy, me not finding blocking issues with a machine assisted PR etc etc ("this code review is provided AS IS and assumes no warranty etc").
